File: Algorithms_4th_Edition/Chapter2/Section3/go_implementation/three_way_quicksort.go
package main

import (
	"algorithms/quicksort/utils"
)

const (
	// N number of trails
	N = 100
	// SIZE size of each array
	SIZE = 10000
)

func main() {
	for i := 0; i < N; i++ {
		arr := utils.UnsortedSlice(SIZE)
		ThreeWayQuickSort(arr)
		utils.CheckIsSorted(arr)
	}
}

// ThreeWayQuickSort : Implementation of three way quick sort method
func ThreeWayQuickSort(arr []int) {
	utils.ShuffleArray(arr)
	_ThreeWayQuickSort(arr)
}

func _ThreeWayQuickSort(arr []int) {
	left := 0
	i := 1
	right := len(arr) - 1
	pivotValue := arr[0]

	for {
		if i > right {
			break
		}

		if arr[i] < pivotValue {
			utils.Swap(arr, i, left)
			i++
			left++
		} else if arr[i] == pivotValue {
			i++
		} else {
			utils.Swap(arr, i, right)
			right--
		}
	}

	leftArraySize := left - 1
	if leftArraySize > 0 {
		ThreeWayQuickSort(arr[0:left])
	}

	rightArraySize := (len(arr) - 1) - (right + 1)
	if rightArraySize > 0 {
		ThreeWayQuickSort(arr[right+1 : len(arr)])
	}
}

File: Algorithms_4th_Edition/Chapter2/Section3/ThreeWayQuickSort.java
package Chapter2.Section3;

import Chapter2.Section1.BaseSort;

/**
 * Quicksort implementation with three way partition.
 * Three way partition method is efficient if there are lots of duplicate items in array.
 */
public class ThreeWayQuickSort extends BaseSort implements ISorter {
    public static void main(String[] args) {
        ThreeWayQuickSort sort = new ThreeWayQuickSort();
        for (int i = 0; i < 100; i++)
            sort.sort(unsortedArray(10000));
        sort.sort(new Comparable[] { 0 });
        sort.sort(new Comparable[] {});
        sort.sort(new Comparable[] { 1, 1, 1, 1, 1, 1, 1, 1, 2 });
    }

    public void sort(Comparable[] a) {
        sort(a, 0, a.length - 1);
        assert isSorted(a);
    }

    public static void sort(Comparable[] a, int lo, int hi) {
        if (lo >= hi)
            return;

        Comparable pivotValue = a[lo];
        int left = lo, i = lo + 1, right = hi;

        while (i <= right) {
            int cmp = a[i].compareTo(pivotValue);
            if (cmp < 0)
                swap(a, i++, left++);
            else if (cmp > 0)
                swap(a, i, right--);
            else
                i++;
        }

        sort(a, lo, left - 1);
        sort(a, right + 1, hi);
    }
}